# Data Understanding

Nesta etapa o objetivo é entender profundamente os dados disponíveis,
avaliando sua estrutura, qualidade, volume e possíveis limitações.

Não há, neste momento, qualquer tipo de tratamento ou limpeza dos dados.
O foco é responder perguntas como:
- Quais dados estão disponíveis?
- Como eles estão estruturados?
- Existem valores ausentes?
- Há dados sensíveis?
- O que pode impactar as análises futuras?

In [1]:
import numpy as np
import pandas as pd
from pathlib import Path

In [5]:
# Definindo quantas colunas o pandas irá exibir
pd.set_option("display.max_columns", 100)

# Definindo como números float serão exibidos (2 casas decimais)
pd.set_option("display.float_format", lambda x: f"{x:,.2f}")

- Lendo os arquivos CSV

In [3]:
DATA_RAW_PATH = Path("../data/raw")

csv_files = sorted(DATA_RAW_PATH.glob("*.csv"))

len(csv_files), csv_files

(12,
 [WindowsPath('../data/raw/abr2025.csv'),
  WindowsPath('../data/raw/ago2025.csv'),
  WindowsPath('../data/raw/dez2025.csv'),
  WindowsPath('../data/raw/fev2025.csv'),
  WindowsPath('../data/raw/jan2025.csv'),
  WindowsPath('../data/raw/jul2025.csv'),
  WindowsPath('../data/raw/jun2025.csv'),
  WindowsPath('../data/raw/mai2025.csv'),
  WindowsPath('../data/raw/mar2025.csv'),
  WindowsPath('../data/raw/nov2025.csv'),
  WindowsPath('../data/raw/out2025.csv'),
  WindowsPath('../data/raw/set2025.csv')])

In [4]:
dfs = []

for file in csv_files:
    df = pd.read_csv(file, encoding="latin1", sep=";")
    df["arquivo_origem"] = file.name
    dfs.append(df)

df_raw = pd.concat(dfs, ignore_index=True)

df_raw.shape

(24009, 16)

Após importar todos os arquivos ".csv" e concatenar todos em um único dataframe, ficamos com **24009 linhas** e **16 colunas**.

## Observações sobre codificação dos arquivos

Durante a ingestão dos dados, foi identificado que os arquivos CSV
não utilizam codificação UTF-8, resultando em erros de leitura.

Após análise, foi adotada a codificação `latin1`, comum em dados
oriundos de sistemas legados da administração pública brasileira,
permitindo a correta leitura de caracteres acentuados.


In [5]:
df_raw.head()

Referência                          Nome  \
0  Folha Complementar - Abril  ACACIO DONIZETTI DE OLIVEIRA   
1        Folha Mensal - Abril  ACACIO DONIZETTI DE OLIVEIRA   
2  Folha Complementar - Abril             ADA ESTER ARCHILA   
3        Folha Mensal - Abril             ADA ESTER ARCHILA   
4  Folha Complementar - Abril       ADAO RODRIGUES DE JESUS   

                   Cargo Data Admissão  Tipo de Regime Descontos   Liquido  \
0            MECANICO II    08/04/1991  REGIME PROPRIO       NaN  1.000,00   
1            MECANICO II    08/04/1991  REGIME PROPRIO  1.584,02  3.603,75   
2  TECNICO EM ENFERMAGEM    04/12/2023  REGIME PROPRIO       NaN  1.000,00   
3  TECNICO EM ENFERMAGEM    04/12/2023  REGIME PROPRIO  1.330,84  5.432,23   
4             JARDINEIRO    08/04/1991  REGIME PROPRIO       NaN  1.000,00   

  Data Desligamento Proventos  Contrato  Atividade Nome Atividade  \
0               NaN  1.000,00      1.00        NaN            NaN   
1               NaN  5.187,77      1.00        NaN            NaN   
2               NaN  1.000,00      1.00        NaN            NaN   
3               NaN  6.763,07      1.00        NaN            NaN   
4               NaN  1.000,00      1.00        NaN            NaN   

  Tipo de Contrato  Data Prevista Termino Contrato  Carga Horária (Sem.)  \
0          Efetivo                             NaN                 40.00   
1          Efetivo                             NaN                 40.00   
2          Efetivo                             NaN                 30.00   
3          Efetivo                             NaN                 30.00   
4          Efetivo                             NaN                 40.00   

  arquivo_origem  
0    abr2025.csv  
1    abr2025.csv  
2    abr2025.csv  
3    abr2025.csv  
4    abr2025.csv

In [6]:
df_raw.tail()

Referência                   Nome  \
24004  Folha Complementar - Setembro  ZAIRA ANGELINA ROGADO   
24005        Folha Mensal - Setembro  ZAIRA ANGELINA ROGADO   
24006        Folha Mensal - Setembro          ZILDA MESSINA   
24007  Folha Complementar - Setembro          ZILDA MESSINA   
24008                            NaN                    NaN   

                        Cargo Data Admissão  Tipo de Regime   Descontos  \
24004      AUXILIAR DE CRECHE    09/12/2013  REGIME PROPRIO         NaN   
24005      AUXILIAR DE CRECHE    09/12/2013  REGIME PROPRIO      323,08   
24006  AUXILIAR DE ENFERMAGEM    02/05/1994  REGIME PROPRIO    1.947,27   
24007  AUXILIAR DE ENFERMAGEM    02/05/1994  REGIME PROPRIO         NaN   
24008                     NaN           NaN             NaN  918.286,26   

            Liquido Data Desligamento     Proventos  Contrato  Atividade  \
24004      1.000,00               NaN      1.000,00      1.00        NaN   
24005      2.442,78               NaN      2.765,86      1.00        NaN   
24006      3.576,00               NaN      5.523,27      1.00        NaN   
24007      1.000,00               NaN      1.000,00      1.00        NaN   
24008  4.302.011,08               NaN  5.220.297,34       NaN        NaN   

      Nome Atividade Tipo de Contrato  Data Prevista Termino Contrato  \
24004            NaN          Efetivo                             NaN   
24005            NaN          Efetivo                             NaN   
24006            NaN          Efetivo                             NaN   
24007            NaN          Efetivo                             NaN   
24008            NaN              NaN                             NaN   

       Carga Horária (Sem.) arquivo_origem  
24004                 40.00    set2025.csv  
24005                 40.00    set2025.csv  
24006                 30.00    set2025.csv  
24007                 30.00    set2025.csv  
24008                   NaN    set2025.csv

## Dicionário de dados

- **Referência**: Indica o tipo de pagamento realizado e o mês de referência da folha.

- **Nome**: Nome completo do servidor público.

- **Cargo**: Cargo ocupado pelo servidor no período analisado.

- **Data Admissão**: Data de ingresso do servidor no quadro da prefeitura.

- **Tipo de Regime**: Regime de Previdência Social ao qual o servidor está vinculado.

- **Descontos**: Valor total de descontos aplicados diretamente no holerite do servidor.

- **Liquido**: Valor líquido efetivamente recebido pelo servidor após os descontos.

- **Data Desligamento**: Data de desligamento do servidor, quando aplicável.

- **Proventos**: Valor bruto da remuneração do servidor antes da aplicação dos descontos.

- **Contrato**: Código interno utilizado pelo sistema da prefeitura para identificação contratual.

- **Atividade**: Código interno utilizado para classificar funções ou atividades específicas no sistema municipal.

- **Nome Atividade**: Descrição textual da atividade ou função exercida pelo servidor.

- **Tipo de Contrato**: Modalidade de vínculo do servidor com a prefeitura.

- **Data Prevista Termino Contrato**: Data prevista para o encerramento do contrato, quando aplicável.

- **Carga Horária (Sem.)**: Carga horária semanal de trabalho do servidor, expressa em horas.

Observação: como podemos ver na última linha do dataframe, a última linha de cada mês é a soma total das seguintes colunas: Descontos, Liquido e Proventos. 

- Lendo apenas o cabeçalho de vários arquivos CSV e criando um dicionário que mapeia cada arquivo para o conjunto de colunas que ele possui.

In [7]:
schemas = {
    file.name: set(pd.read_csv(file, encoding="latin1", sep=";", nrows=0).columns)
    for file in csv_files
}

schemas

{'abr2025.csv': {'Atividade',
  'Carga Horária (Sem.)',
  'Cargo',
  'Contrato',
  'Data Admissão',
  'Data Desligamento',
  'Data Prevista Termino Contrato',
  'Descontos',
  'Liquido',
  'Nome',
  'Nome Atividade',
  'Proventos',
  'Referência',
  'Tipo de Contrato',
  'Tipo de Regime'},
 'ago2025.csv': {'Atividade',
  'Carga Horária (Sem.)',
  'Cargo',
  'Contrato',
  'Data Admissão',
  'Data Desligamento',
  'Data Prevista Termino Contrato',
  'Descontos',
  'Liquido',
  'Nome',
  'Nome Atividade',
  'Proventos',
  'Referência',
  'Tipo de Contrato',
  'Tipo de Regime'},
 'dez2025.csv': {'Atividade',
  'Carga Horária (Sem.)',
  'Cargo',
  'Contrato',
  'Data Admissão',
  'Data Desligamento',
  'Data Prevista Termino Contrato',
  'Descontos',
  'Liquido',
  'Nome',
  'Nome Atividade',
  'Proventos',
  'Referência',
  'Tipo de Contrato',
  'Tipo de Regime'},
 'fev2025.csv': {'Atividade',
  'Carga Horária (Sem.)',
  'Cargo',
  'Contrato',
  'Data Admissão',
  'Data Desligamento',
  'D

- Identificando quantas estruturas de colunas diferentes existem entre todos os arquivos CSV.

In [8]:
unique_schemas = set(tuple(sorted(cols)) for cols in schemas.values())
len(unique_schemas)

1

Todos os arquivos têm exatamente o mesmo schema.

- Colunas do dataframe

In [8]:
df_raw.columns

Index(['Referência', 'Nome', 'Cargo', 'Data Admissão', 'Tipo de Regime',
       'Descontos', 'Liquido', 'Data Desligamento', 'Proventos', 'Contrato',
       'Atividade', 'Nome Atividade', 'Tipo de Contrato',
       'Data Prevista Termino Contrato', 'Carga Horária (Sem.)',
       'arquivo_origem'],
      dtype='object')

## Resumo da compreensão dos dados

- Os dados estão distribuídos em múltiplos arquivos CSV, um por período.
- Todos os arquivos apresentam **schema consistente**, permitindo concatenação segura.
- A codificação utilizada é `latin1`, compatível com sistemas legados.
- Não foram identificadas divergências estruturais entre os meses.
- Existem colunas com dados potencialmente sensíveis, que serão tratadas
  nas próximas etapas.
